crates/matrix-sdk/src/room/mod.rs

```diff
@@ -3195,6 +3201,135 @@ impl Room {
         ObservableLiveLocation::new(&self.client, self.room_id())
     }
 
+    /// Helper to requests to join this `Room`.
+    ///
+    /// The current requests to join the room will be emitted immediately
+    /// when subscribing. When a new membership event is received, a request is
+    /// marked as seen or there is a limited sync, a new set of requests
+    /// will be emitted.
+    pub async fn subscribe_to_join_requests(&self) -> Result<impl Stream<Item = Vec<JoinRequest>>> {
+        let this = Arc::new(self.clone());
+
+        let requests_observable =
+            this.client.observe_room_events::<SyncRoomMemberEvent, (Client, Room)>(this.room_id());
+
+        let (current_seen_ids, mut seen_request_ids_stream) =
+            this.subscribe_to_seen_join_request_ids().await?;
+
+        let mut room_info_stream = this.subscribe_info();
+
+        let combined_stream = stream! {
+            // Emit current requests to join
+            match this.clone().get_current_join_requests(&current_seen_ids).await {
+                Ok(initial_requests) => yield initial_requests,
+                Err(e) => warn!("Failed to get initial requests to join: {e:?}")
+            }
+
+            let mut requests_stream = requests_observable.subscribe();
+
+            let mut new_event: Option<SyncRoomMemberEvent> = None;
+            let mut seen_ids = current_seen_ids.clone();
+            let mut prev_seen_ids = current_seen_ids;
+            let mut prev_missing_room_members: bool = false;
+            let mut missing_room_members: bool = false;
+
+            loop {
+                // This is equivalent to a combine stream operation, triggering a new emission
+                // when any of the branches changes
+                tokio::select! {
+                    Some((next, _)) = requests_stream.next() => { new_event = Some(next); }
+                    Some(next) = seen_request_ids_stream.next() => { seen_ids = next; }
+                    Some(next) = room_info_stream.next() => {
+                        missing_room_members = !next.are_members_synced()
+                    }
```

Comment:
So, if I'm reading the code correctly: it seems the `select!` tries to do the minimal amount of work, and then the actual work is done a bit after. Another way would have been to essentially take the code that reacts to a change here, and put it immediately in the branches. Any reason why it was not done this way? It would preclude the need for a few local variables, and make reading the code a bit more straightforward (instead of `if this or this or this happened, consider that, and maybe do this or that`, we'd have `if this happened, do that; if thot happened, do thut` etc.)